Replace gossip follower with dedicated node sync follower

mystenmark opened the pull request:
This code accomplishes the following:
- Removes dependency on gossip follower code, which has a lot of behavior that interacts poorly with the fullnode case, specifically the timeouts and reconnect logic.
- Stores all follower items to a database immediately, and tries much harder not to refetch portions of the batch stream multiple times.
- Adds exponential backoff when the follower encounters errors.
- Removes node-follow-specific code from gossip/mod.rs

Note that node_state.rs was moved - there is no new code there even though github says its all new.

mystenmark commented on crates/sui-core/src/node_sync/node_state.rs:
```diff
@@ -0,0 +1,779 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
```
This file was moved, unchanged - this code is not new.

longbowlu commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,535 @@
-            })?;
-        } else {
-            trace!("no pending waiters");
+        for (name, _) in aggregator.committee.members() {
```
just to double note that in the old logic, one FN follows `|committee| - 1` validators but here it is changed. There are pros and cons (cons especially when the committee size is large) and we may need to revisit it in the future

mystenmark replied:
The old code subtracts one because it assumes that it is a validator, and it doesn't want to follow itself.

Fullnodes should absolutely follow the entire committee - they require finality, and the only way to be sure that you will discover finality is to follow the entire committee. (You need 2f+1 to attest to the cert. Any subset of the committee may contain up to f byzantine nodes. Therefore there is no proper subset that is guaranteed to contain 2f+1 honest nodes).

longbowlu replied:
following all validators to get quorum assumes the worst case, which we should certainly prepare for. But in majority cases, we will not have f faulty nodes. Even if we have a handful (<f), this can be fixed by connecting with different nodes in the next round (similarly to how do we it in the current code where we randomly choose many but not all of the validators). 

The trade-off is load on validators, maybe we need to worry about it after this PR, or checkpoint fully works, but sth to keep in mind

mystenmark replied:
The solution to the scalability problems is to not have every fullnode in the world connect directly to the validators. We shouldn't compromise liveness here in order to put a bandaid on a larger problem.

longbowlu commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,535 @@
+            // At least 3 more were fetched from the stream - this number can't be checked exactly
+            // because the authority returns everything in the current batch even if fewer items
+            // were requested, and batch boundaries are non-deterministic.
+            assert!(result.items_from_stream >= 3 && result.items_from_stream <= 5);
```
the batch size is 5? any reasons to keep it this small?

mystenmark replied:
No, the batch size can be very large, but in low throughput situations new batches are created due to timers, not to exceeding a threshold. So in this case the size of the batch is determined by when the timer fires.

longbowlu replied:
but how do we know for sure it's <= 5?

mystenmark replied:
empirically, this is what made the test pass. It can never be less than 3, because the test handler won't emit a failure until it sees at least 3. The actual number can be a bit larger than 3, but the important thing is that we fail if it is 10, because that might indicate that the failure was not injected.

I recognize this is not the prettiest test code in the world, but this stuff is very difficult to test and I already spent most of a full day on these tests just to get them to the point they are at.

longbowlu commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,535 @@
 
-        let digests = TransactionEffectsDigest::random();
+            // At least 3 more were fetched from the stream - this number can't be checked exactly
+            // because the authority returns everything in the current batch even if fewer items
+            // were requested, and batch boundaries are non-deterministic.
```
interesting, why is this better than always starting from the requested seq number, and why sin't batch boundary deterministic?

longbowlu commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,535 @@
-        // double vote doesn't result in false if we already exceeded threshold.
-        assert!(map.note_effects_digest(&byzantine, 1, threshold, &digests));
+            let ivs = intervals.lock().unwrap();
+            assert_eq!(ivs.len(), 7);
+            assert!(*ivs.last().unwrap() > Duration::from_secs(30));
```
it's not very self-descriptive why `7` and `30`. let's add more comments if people can easily understand what may go wrong.

mystenmark replied:
commented

longbowlu reviewed (approved):
Great work! It's now much more elegant than the gossip process.

mystenmark commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,535 @@
-        }
+                    Some(Err(err)) => {
+                        debug!(?peer, "handle_batch_stream error: {}", err);
+                        return Err(err);
```
We could, but the way this is designed, it shouldn't matter very much, and this keeps the code simpler.

We will restart anything that was in progress, and typically we won't repeat any work. The sequence will be read from the db instead of from the validator, and if the sync process had previously made any progress on other certs (e.g. downloading certs/effects) then nothing will be re-downloaded.

mystenmark commented on crates/sui-node/src/lib.rs:
```diff
@@ -222,11 +214,8 @@ impl SuiNode {
-                        None,
-                        None,
-                    )
+                    active_authority.respawn_node_sync_process().await;
+                    (None, None, None)
```
I don't think I understand what you want me to change here - fullnode always starts sync process in the current code.

longbowlu replied:
The logic does not change, I'm suggesting we simplify the winding code

mystenmark replied:
sure - i tried to, and it doesn't seem to actually get much simpler. I can do this: https://github.com/MystenLabs/sui/commit/5f931e75eac1ecdde98561c6a7f04828ff8a0428 - but its not obvious if it can be made a ton simpler than that without major refactoring.

lxfind replied:
I (somewhat) cleaned this up a tiny bit in https://github.com/MystenLabs/sui/pull/4233

mystenmark commented on crates/sui-core/src/authority_active.rs:
```diff
@@ -333,16 +333,47 @@ where
-        // eventually find finality, therefore we must follow all validators.
-        let target_num_tasks = committee.num_members();
+    pub async fn respawn_node_sync_process(&self) {
+        info!("respawn_node_sync_process");
```
Printed out the epoch number. The full committee info should probably be logged during reconfiguration.

laura-makdah reviewed (changes requested):
I know that this code solves pressing issues so I am hesitant block this change arguing against a divergence, but the code in node_follower.rs looks to me like an almost exact re-write of the code in gossip/mod.rs 

I am concerned about the operational complexity of maintaining two versions code that do 90% of the same thing, especially when considering that this functionality is expected to go through several more iterations as we scale up, replace the network to p2p, implement a different communication topology, integrate sparse fullnodes, and fix bugs. I am worried especially that the bugs in fullnodes will be more apparent since the scale of the number of fullnodes is much larger, but fixes will not make it to the original validator gossip code even when the issue also exists there. 

I understand that it is challenging to handle slight differences in functionality cleanly within the same loop, but after seeing this divergent version I would argue this is going to cause way more operational complexity and issues in the long run.  I think a lot of these changes are great, a big part of my argument for keeping the processes shared is that I want to see these great improvements and cleanups in the validator gossip as well.

mystenmark commented:
> I am concerned about the operational complexity of maintaining two versions code that do 90% of the same thing, especially when considering that this functionality is expected to go through several more iterations as we scale up, replace the network to p2p, implement a different communication topology, integrate sparse fullnodes, and fix bugs. I am worried especially that the bugs in fullnodes will be more apparent since the scale of the number of fullnodes is much larger, but fixes will not make it to the original validator gossip code even when the issue also exists there.

I originally came at this from your perspective (don't duplicate code), but I think that was a mistake, and this is 100% the correct long term solution. Gossip behaves very differently from node following, and trying to re-use the same code for both was a mistake. As evidence of this, you can see all the special cases I've removed from the gossip code that we put in to try and deal with these differences. To completely fix the fullnode issues, nearly ever aspect of gossip would have to have two different behaviors, one for fullnode, and one for gossip. Here's a partial list:

- Gossip follows a subset, node follows entire committee.
- Gossip times out and restarts connections periodically, node has no need to do that.
- Gossip doesn't want to re-follow the same peer immediately after disconnecting, node does not.
- Node wants to pick up from where it left off when reconnecting - Gossip should not do this.
- Nodes may create thundering herd problems, while gossip can't, hence full node should have exponential backoff.

It's possible that there may some bugs that have to be fixed twice because of this duplication. However, I think continuing to share code creates a much worse situation in which we attempt to fix a bug that affects nodes and end up introducing a gossip bug (or vice versa). In fact this has already happened at least once: When I added the follower_db, we ended up introducing unwanted "pick up where i left off" behavior to gossip.

mystenmark commented on crates/sui-storage/src/node_sync_store.rs:
```diff
@@ -43,4 +52,56 @@ impl NodeSyncStore {
+        let mut write_batch = self.batch_streams.batch();
+        trace!(?peer, ?seq, ?digests, "persisting digests to db");
+        write_batch = write_batch
+            .insert_batch(&self.batch_streams, std::iter::once(((peer, seq), digests)))?;
```
The protection is that the write is idempotent - do you envision a need for something more than that?

laura-makdah replied:
The goal of the changes here is to reduce unnecessary work right, so does anything ensure this isn't being called multiple times?

mystenmark commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,538 @@
-    sync_arg: SyncArg,
-    tx: Option<oneshot::Sender<SuiResult>>,
-}
+        loop {
```
I'm confused - this code does have sleeps and delays. (Search for `MIN_RECONNECTION_DELAY`). Can you explain the scenario you are worried about?

laura-makdah replied:
the `MIN_RECONNECTION_DELAY ` is not sufficient, We need a sleep statement in both the outer loop and the inner loop that is guaranteed to be hit at least every once in a while, because these are both infinite loops. Without sleep statements, once a tread starts execution of the loop, it will never be released back to the runtime.

mystenmark commented on crates/sui-core/src/node_sync/node_follower.rs:
```diff
@@ -2,797 +2,538 @@
+            let digests = $digests;
+            let fut = sync_handle.handle_digest(peer, seq, digests).await?;
+            results.push(result_block(fut, seq, digests));
+        }};
```
this invokes `handle_digest` and pushes the result future into the `results` queue.

I'm sorry about the macro but I couldn't find a way to avoid duplicating this code with functions, due to the borrow checker.

mystenmark commented on crates/sui-core/src/authority_active/gossip/mod.rs:
```diff
@@ -402,40 +371,13 @@ impl<A> Follower<A>
-            GossipType::Full => Some(start_seq),
-        };
-
-        debug!(peer = ?peer_name, ?start_seq, "Restarting follower at sequence");
```
My reading is that in main, for `GossipType::BestEffort`, `max_seq` never contains any value other than `None`, which is why I removed it.

This, imo, points towards the need to separate these two pieces of code. The gossip code is so polluted with special-cases for nodes that neither case does what we actually want it to.

mystenmark commented on crates/sui-core/src/authority_active.rs:
```diff
@@ -108,13 +114,15 @@ impl AuthorityHealth {
     pub state: Arc<AuthorityState>,
     pub node_sync_state: Arc<NodeSyncState<A>>,
     node_sync_handle: OnceCell<NodeSyncHandle>,
+    node_sync_process: Arc<Mutex<Option<NodeSyncProcessHandle>>>,
```
Very good point, added comments.

mystenmark commented on crates/sui-core/src/authority_active/gossip/mod.rs:
```diff
@@ -38,21 +33,10 @@ mod configurable_batch_action_client;
     client: SafeClient<A>,
     state: Arc<AuthorityState>,
-    follower_store: Arc<FollowerStore>,
-    max_seq: Option<TxSequenceNumber>,
```
See below.

laura-makdah commented:
> I originally came at this from your perspective (don't duplicate code), but I think that was a mistake, and this is 100% the correct long term solution. Gossip behaves very differently from node following, and trying to re-use the same code for both was a mistake. As evidence of this, you can see all the special cases I've removed from the gossip code that we put in to try and deal with these differences. To completely fix the fullnode issues, nearly ever aspect of gossip would have to have two different behaviors, one for fullnode, and one for gossip. Here's a partial list:

There were mistakes in the code that caused a lot of issues, but doing a full re-write isn't the only way to fix mistakes, and it seems unnecessary to start from scratch. 

> 
> * Gossip follows a subset, node follows entire committee.

This topology is expected to change as we scale up, and can be easily specified a configuration file.

> * Gossip times out and restarts connections periodically, node has no need to do that.

When the topology changes as we all expect it must in order to scale up, node will also need the ability to restart and rotate connections. For now this can be handled again by a config variable that specifies the timeout duration, and for full nodes set to the largest possible value. 

> * Gossip doesn't want to re-follow the same peer immediately after disconnecting, node does not.

When the topology changes, nodes will not want to immediately re-follow the same peer either. For now, choosing peers to connect to can be done via separate functions, along with a config variable.

> * Node wants to pick up from where it left off when reconnecting - Gossip should not do this.

This is the only nontrivial difference I see out of all these that requires a bit of extra code. But this alone doesn’t warrant having two separate processes that each need to be continually maintained. 

> * Nodes may create thundering herd problems, while gossip can't, hence full node should have exponential backoff.

If we solve the herd problem in full nodes, validators would also benefit from that improvement, even though it their scaling problems are not as imminent. 

> It's possible that there may some bugs that have to be fixed twice because of this duplication. However, I think continuing to share code creates a much worse situation in which we attempt to fix a bug that affects nodes and end up introducing a gossip bug (or vice versa). In fact this has already happened at least once: When I added the follower_db, we ended up introducing unwanted "pick up where i left off" behavior to gossip.

I think using a config and keeping this in mind going forward will minimize issues. I've seen it happen at other companies where people solve the same problem several times in slightly different ways because they want ownership of something. People who duplicated a solution for themself can iterate on it can then quicker without getting buy-in from others, but then overall as a company, a lot of effort is duplicated on solutions that are duplicated. I don't think that is your intention, but would rather that doesn't happen here.

mystenmark commented:
@laura-makdah I have spent the last 6-8 weeks pulling my hair out trying to fix the issues with fullnode **without** duplicating code, and I do not believe that it is a tenable approach.

mystenmark commented on crates/sui-storage/src/node_sync_store.rs:
```diff
@@ -12,12 +13,20 @@ use typed_store::traits::DBMapTableUtil;
+    batch_streams: DBMap<(AuthorityName, TxSequenceNumber), ExecutionDigests>,
+
+    /// The latest received sequence from each authority.
+    latest_seq: DBMap<AuthorityName, TxSequenceNumber>,
```
I tried that first - the problem is that `batch_streams` can be emptied once we've processed everything in it, and then we've forgotten where we were.

mystenmark commented on crates/sui-core/src/authority_active.rs:
```diff
@@ -333,16 +339,47 @@ where
+                .is_err()
+            {
+                info!("aborting node sync task");
+                join_handle.abort();
```
Good point re joining.

I try to reserve error for things that "can't" happen. IMO every `ERROR` level log we get in production should indicate a bug that requires investigation. (This is just a personal preference and not a policy that we've adopted collectively).

andll replied:
That sort of should not happen right? The way code is structured(with high level select on the guard) task not completing likely mean something bad / highly unexpected like deadlock or some other degradation. Perhaps `warn` then?

mystenmark replied:
Ok, yes, you're right - it definitely should not happen.
